exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java

```diff
@@ -158,38 +165,49 @@ public DrillbitContext getContext() {
     return dContext;
   }
 
-  private ExtendedLatch exitLatch = null; // used to wait to exit when things are still running
-
   /**
    * Waits until it is safe to exit. Blocks until all currently running fragments have completed.
-   *
-   * <p>This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.</p>
+   * This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.
    */
   public void waitToExit() {
-    synchronized(this) {
-      if (queries.isEmpty() && runningFragments.isEmpty()) {
-        return;
+    final long startTime = System.currentTimeMillis();
+    final long endTime = startTime + EXIT_TIMEOUT;
+
+    exitLock.lock();
+
+    try {
+      long currentTime;
+      while ((currentTime = System.currentTimeMillis()) < endTime) {
+        if (queries.isEmpty() && runningFragments.isEmpty()) {
+          break;
+        }
+
+        try {
+          if (!exitCondition.await(endTime - currentTime, TimeUnit.MILLISECONDS)) {
+            break;
+          }
+        } catch (InterruptedException e) {
+          logger.error("Interrupted while waiting to exit");
+        }
       }
 
-      exitLatch = new ExtendedLatch();
-    }
+      if (!(queries.isEmpty() && runningFragments.isEmpty())) {
+        logger.warn("Timed out after {} millis. Shutting down before all fragments and foremen " +
```

Comment:
It may be good to add what queries and runningFragments are running to the log message.

Reply:
Done